commit 6456850a46ab40923206cc3952d29ab375361604
Author: Andreas Oberritter <obi@linuxtv.org>
Date:   Mon May 9 00:03:34 2011 +0100

    [media] DVB: return meaningful error codes in dvb_frontend
    
    - Return values should not be ORed. Abort early instead.
    - Return -EINVAL instead of -1.
    
    Signed-off-by: Andreas Oberritter <obi@linuxtv.org>
    Signed-off-by: Mauro Carvalho Chehab <mchehab@redhat.com>

--- a/drivers/media/dvb/dvb-core/dvb_frontend.c	2011-08-09 18:26:58.288244025 +0200
+++ b/drivers/media/dvb/dvb-core/dvb_frontend.c	2011-08-09 18:41:03.576233778 +0200
@@ -1337,12 +1337,12 @@
 		tvp->u.data = fe->dtv_property_cache.isdbs_ts_id;
 		break;
 	default:
-		r = -1;
+		return -EINVAL;
 	}
 
 	dtv_property_dump(tvp);
 
-	return r;
+	return 0;
 }
 
 static int dtv_property_process_set(struct dvb_frontend *fe,
@@ -1355,11 +1355,11 @@
 	dtv_property_dump(tvp);
 
 	/* Allow the frontend to validate incoming properties */
-	if (fe->ops.set_property)
+	if (fe->ops.set_property) {
 		r = fe->ops.set_property(fe, tvp);
-
-	if (r < 0)
-		return r;
+		if (r < 0)
+			return r;
+	}
 
 	switch(tvp->cmd) {
 	case DTV_CLEAR:
@@ -1378,7 +1378,7 @@
 		dprintk("%s() Finalised property cache\n", __func__);
 		dtv_property_cache_submit(fe);
 
-		r |= dvb_frontend_ioctl_legacy(inode, file, FE_SET_FRONTEND,
+		r = dvb_frontend_ioctl_legacy(inode, file, FE_SET_FRONTEND,
 			&fepriv->parameters);
 		break;
 	case DTV_FREQUENCY:
@@ -1493,7 +1493,7 @@
 		fe->dtv_property_cache.isdbs_ts_id = tvp->u.data;
 		break;
 	default:
-		r = -1;
+		return -EINVAL;
 	}
 
 	return r;
@@ -1567,8 +1567,10 @@
 		}
 
 		for (i = 0; i < tvps->num; i++) {
-			(tvp + i)->result = dtv_property_process_set(fe, tvp + i, inode, file);
-			err |= (tvp + i)->result;
+			err = dtv_property_process_set(fe, tvp + i, inode, file);
+			if (err < 0)
+				goto out;
+			(tvp+i)->result = err;
 		}
 
 		if(fe->dtv_property_cache.state == DTV_TUNE)
@@ -1599,8 +1601,10 @@
 		}
 
 		for (i = 0; i < tvps->num; i++) {
-			(tvp + i)->result = dtv_property_process_get(fe, tvp + i, inode, file);
-			err |= (tvp + i)->result;
+			err = dtv_property_process_get(fe, tvp + i, inode, file);
+			if (err < 0)
+				goto out;
+			(tvp + i)->result = err;
 		}
 
 		if (copy_to_user(tvps->props, tvp, tvps->num * sizeof(struct dtv_property))) {
